File: backend/Link.js
export default class Link {
  constructor(accounts, validFrom, signatures = {}) {
    this.accounts = accounts;
    this.validFrom = validFrom;
    this.signatures = signatures;
    if (
      Object.keys(this.signatures).some(
        publicKey =>
          !this.accounts.map(account => account.publicKey).includes(publicKey)
      )
    )
      throw new Error("all signing accounts must be part of link");
  }

  addSignature(account, signature, verify = true) {
    if (!(account in this.accounts))
      throw new Error("signing account must be part of link");
    if (verify) {
      if (!account.key.verify(this.toSign, signature))
        throw new Error("invalid signature!");
    }
    if (this.signatures[account.key.publicKey])
      throw new Error("link already signed by account");
    this.signatures[account.key.publicKey] = signature;
  }

  /** Assumes validity of signatures, only checks if signed by all parties and not expired. */
  get valid() {
    if (Object.keys(this.signatures).length !== this.accounts.length)
      return false;
    return Date.now() < this.validUntil;
  }

  get validUntil() {
    return this.validFrom + validityPeriod;
  }

  /** String that should be signed by accounts, the same for everyone. */
  get toSign() {
    const accountNames = this.accounts.map(account => account.name).sort();
    const publicKeys = this.accounts.map(account => account.name).sort();
    return [accountNames.join(":"), publicKeys.join(":"), this.validFrom].join(
      "&"
    );
  }
}

const validityPeriod = 30 * 24 * 60 * 60 * 1000;

File: lib/network/link.dart
import 'package:liberum/network/account.dart';

class Link {
  List<Account> accounts;
  List<String> signatures;

  Link(this.accounts, this.signatures, this._validFrom, {bool verify=true}) {
    if (this.accounts.length != this.signatures.length) {
      throw ArgumentError('Accounts and signatures lists must have the same length');
    }
    if (verify) {
      for (int i = 0; i < this.accounts.length; i++) {
        if (!this.accounts[i].key.verify(this.signable, this.signatures[i])) {
          throw new ArgumentError('Forged signature');
        }
      }
    }
  }

  DateTime _validFrom;
  DateTime get validFrom => _validFrom;

  DateTime get validUntil => validFrom.add(Duration(days: 30));

  bool get isValid => DateTime.now().isBefore(this.validUntil);


  String get signable => _accountStrings + '|' + _validFrom.hashCode.toRadixString(16);
  String get _accountStrings => accounts.map((account) => account.key.publicFingerprint).join('|');
}